http-transport-binding.md

```diff
@@ -158,28 +158,28 @@ are individually mapped to and from distinct HTTP message headers.
 
 ##### 3.1.3.1 HTTP Header Names
 
-The naming convention for the HTTP header mapping of attributes is:
-
-    * Each attribute name MUST be prefixed with "CE-"
+Except for attributes [explicitly handled in this specification]
+(#2-use-of-cloudevents-attributes), the naming convention for the 
+HTTP header mapping of well-known CloudEvents attributes is that 
+each attribute name MUST be prefixed with "CE-".
 
 Examples:
 
     * `eventTime` maps to `CE-EventTime`
     * `eventID` maps to `CE-EventID`
     * `cloudEventsVersion` maps to `CE-CloudEventsVersion`
 
-For the `extensions` attribute, each entry of the `extensions` map
-is mapped to a separate HTTP header. The `extensions` attribute itself is
-not mapped to a header.
-
-The naming convention for the `extensions` header mapping of attributes is:
-
-    * Each map entry name MUST be prefixed with "CE-X-"
-
-Examples:
-
-    * `example` maps to `CE-X-Example`
-    * `testExtension` maps to `CE-X-TestExtension`
+`Map`-typed CloudEvents attributes MUST be flattened into a set
+of HTTP headers, where by the name of each header carries the prefix
+"CE-", an infix reflecting the map attribute followed by a dash 
+("-"), and the name of the map entry key, e.g. "CE-attrib-key".
+
+CloudEvents extensions that define their own attributes MAY define a 
+diverging mapping to HTTP headers for those attributes, especially if 
```

Comment:
What happens in a situation where `producer` and `consumer` understand the extension but `middleware` does not?

` producer -> middleware -> consumer`

Does this PR allow the extension to define its HTTP header to be something other than a `CE-` prefixed extension header? I'm trying to understand how `middleware` would know to transmit the data the consumer. If we just send all the original headers to `consumer` this might work, but how do we disambiguate between some other headers that are not related to CloudEvents? What if the consumer expects the JSON form of CloudEvents?

Reply:
@clemensv ^^

My 2 cents... I think the middleware would have to forward all headers on to the next hop.

re: what if the next hop expects JSON.... this is why the spec says everyone SHOULD support both so its less of an issue. However, I've been thinking more about @inlined's suggestion that we drop the Structured version and just support Binary HTTP - it's sounding more appealing each time I think about it. While I do still value in defining a JSON serialization of CloudEvents (if nothing else so we have interop for how they look when written in JSON), by forcing all CE properties into HTTP headers I think people would be less inclined to treat them as extensions of `data` and put a ton of non-routing specific info into them. It might help ensure we keep things simple. But still thinking about it.

Reply:
HTTP has clear rules about this for its intermediaries. Proxies must forward all RFC7231 headers and that includes all extensions. The same is true for MQTT and AMQP. For the transport bindings that makes the raised point implicitly clear for within a transport, which what they subject of this PR is.

I am not in favor of dropping the structured format, but I am sympathetic to making the header mapping a MUST rule for the structured format rather than a SHOULD, i.e. all metadata headers and extensions MUST be visible to middleware at all times.

The case where the structured format has a clear advantage is for multi-hop routing, when an event gets forwarded HTTP->AMQP->AMQP->[Avro file on disk]->HTTP. The structured format keeps everything together. 

In the CloudEvents main spec we should indeed discuss routing/forwarding explicitly and make a hard rule that middleware MUST forward all elements of an event, including those that have been mapped onto transport constructs.

Reply:
I assume this can be done in a follow-on PR since that would be additive and not a change to this PR - plus that change is independent of the purpose of this PR too.